### PR TITLE
voice receive support

### PR DESCRIPTION
**Pull request by cakedan**
creates a VoiceData event and a VoiceSpeaking event

I didn't send in a user object with either event, just a user_id, didn't know if you wanted the full user object to be sent with them or not

### CODE REVIEW

File: disco/voice/client.py [hunk 1]
```diff
@@ -80,6 +102,10 @@ def __init__(self, channel, encoder=None, max_reconnects=5):
         self._update_listener = None
         self._heartbeat_task = None
 
+        # SSRCs
+
```
Comment (b1naryth1ef) [inline disco/voice/client.py]: drop this newline

File: disco/voice/client.py [hunk 2]
```diff
@@ -144,6 +186,19 @@ def on_voice_ready(self, data):
             self.disconnect()
             return
 
+        codecs = []
+
+        for i in range(len(AudioCodecs)):
```
Comment (b1naryth1ef) [inline disco/voice/client.py]:
```
for idx, codec in enumerate(AudioCodecs):
```

File: disco/voice/client.py [hunk 3]
```diff
@@ -144,6 +186,19 @@ def on_voice_ready(self, data):
...
+            codec = AudioCodecs[i]
+            payload_type = PayloadTypes.get(codec)
+            if payload_type:
+                codecs.append({
```
Comment (b1naryth1ef) [inline disco/voice/client.py]: Comment describing whats going on here would be good.

File: disco/voice/udp.py [hunk 4]
```diff
@@ -35,9 +66,9 @@ def __init__(self, vc):
         self._secret_box = None
 
         # Buffer used for encoding/sending frames
-        self._buffer = bytearray(24)
-        self._buffer[0] = 0x80
-        self._buffer[1] = 0x78
+        self._header = bytearray(12)
+        self._header[0] = 2 << 6  # Only RTP Version set in the first byte of the header, 0x80
```
Comment (b1naryth1ef) [inline disco/voice/udp.py]: Raise this to a constant

File: disco/voice/udp.py [hunk 5]
```diff
@@ -85,7 +129,105 @@ def send_frame(self, frame, sequence=None, timestamp=None, incr_timestamp=None):
 
     def run(self):
         while True:
-            self.conn.recvfrom(4096)
+            data, addr = self.conn.recvfrom(4096)
+
+            # Data cannot be less than the bare minimum, just ignore
+            if len(data) <= 12:
```
Comment (b1naryth1ef) [inline disco/voice/udp.py]: Throw a logging debug statement in here, will be useful in some scenarios.

File: disco/voice/udp.py [hunk 6]
```diff
@@ -85,7 +129,105 @@ def send_frame(self, frame, sequence=None, timestamp=None, incr_timestamp=None):
...
+                csrc_count=first & 0x0F,
+                marker=second >> 7,
+                payload_type=second & 0x7F,
+                sequence=sequence,
+                timestamp=timestamp,
+                ssrc=ssrc,
+            )
+
+            # Check if rtp version is 2
+            if rtp.version != 2:
```
Comment (b1naryth1ef) [inline disco/voice/udp.py]: Same in here, basically anyplace we ignore/throw away the packet in this function we should throw a debug log.

**Comment by Azoy** [comment]
🎉

**Pull request merged**